In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [42]:
# Load datasets
customers = pd.read_csv("olist_customers_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
payments = pd.read_csv("olist_order_payments_dataset.csv")
reviews = pd.read_csv("olist_order_reviews_dataset.csv")
products = pd.read_csv("olist_products_dataset.csv")
sellers = pd.read_csv("olist_sellers_dataset.csv")
geolocation = pd.read_csv("olist_geolocation_dataset.csv")
category_translation = pd.read_csv("product_category_name_translation.csv")

In [43]:
datasets = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

summary = []

for name, df in datasets.items():
    summary.append({
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1],
        "Duplicate Rows": df.duplicated().sum(),
        "Total Missing Values": df.isnull().sum().sum()
    })

summary_df = pd.DataFrame(summary)

summary_df

,Dataset,Rows,Columns,Duplicate Rows,Total Missing Values
0,customers,99441,5,0,0
1,orders,99441,8,0,4908
2,order_items,112650,7,0,0
3,payments,103886,5,0,0
4,reviews,99224,7,0,145903
5,products,32951,9,0,2448
6,sellers,3095,4,0,0
7,geolocation,1000163,5,261831,0
8,category_translation,71,2,0,0


In [44]:
def profile_dataset(df, name):
    print("=" * 60)
    print(f"DATASET: {name}")
    print("=" * 60)

    print(f"\nRows: {df.shape[0]:,}")
    print(f"Columns: {df.shape[1]}")

    print("\nColumn Names:")
    print(df.columns.tolist())

    print("\nMissing Values:")
    print(df.isnull().sum())

    print("\nDuplicate Rows:")
    print(df.duplicated().sum())

    print("\nData Types:")
    print(df.dtypes)

    print("\nSample Records:")
    print(df.head())

    print("\n")

In [45]:
datasets = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

for name, df in datasets.items():
    profile_dataset(df, name)

DATASET: customers

Rows: 99,441
Columns: 5

Column Names:
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

Missing Values:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Duplicate Rows:
0

Data Types:
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

Sample Records:
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c7

In [46]:
datasets = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

summary = []

for name, df in datasets.items():
    summary.append({
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1],
        "Duplicate Rows": df.duplicated().sum(),
        "Total Missing Values": df.isnull().sum().sum()
    })

summary_df = pd.DataFrame(summary)

summary_df

,Dataset,Rows,Columns,Duplicate Rows,Total Missing Values
0,customers,99441,5,0,0
1,orders,99441,8,0,4908
2,order_items,112650,7,0,0
3,payments,103886,5,0,0
4,reviews,99224,7,0,145903
5,products,32951,9,0,2448
6,sellers,3095,4,0,0
7,geolocation,1000163,5,261831,0
8,category_translation,71,2,0,0


In [47]:
customers["customer_id"].nunique()


99441

In [48]:
len(customers)


99441

In [49]:
orders["order_id"].nunique()

99441

In [50]:
len(orders)

99441

In [51]:
products["product_id"].nunique()

32951

In [52]:
len(products)


32951

In [53]:
sellers["seller_id"].nunique()

3095

In [54]:
len(sellers)

3095

In [55]:
reviews["review_id"].nunique()

98410

In [56]:
len(reviews)

99224

In [57]:
reviews[reviews["review_id"].duplicated(keep=False)] \
[['review_id','order_id']] \
.sort_values('review_id') \
.head(20)

,review_id,order_id
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9


### Customers Table - Key Observations

* The customers dataset contains 99,441 records and 5 columns.
* No missing values or duplicate rows were identified.
* customer_id appears suitable as the primary business key.
* Customer location information is available at city and state level, enabling future geographic analysis.
* Overall, the dataset is clean and requires no preprocessing.


### Orders Table - Key Observations

* The orders dataset contains 99,441 records and captures the complete order lifecycle.
* Missing values exist in approval and delivery-related timestamps.
* These missing timestamps likely correspond to cancelled, unavailable, or undelivered orders rather than data quality issues.
* No duplicate records were found.
* The table will be critical for delivery performance and order fulfillment analysis.


### Remaining Tables - Key Observations

**Order Items**

* Contains 112,650 records with no missing values or duplicates.
* Represents product-level transactions and is a strong candidate for the fact table grain.

**Payments**

* Contains 103,886 payment records with complete data quality.
* Supports payment behavior and installment analysis.

**Reviews**

* Review scores are complete, but review titles and comments contain substantial missing values.
* Missing comments are expected because textual feedback is optional.
* review_id should not automatically be assumed to be unique and requires validation.

**Products**

* Most product records are complete.
* 610 products are missing category-related metadata and 2 products have missing dimensions.
* Missing values affect only a small proportion of records.

**Sellers**

* Clean dataset with no missing values or duplicates.
* Suitable for seller-level performance analysis.

**Geolocation**

* Contains over 1 million records.
* Significant duplicate records (261,831) were detected.
* May require deduplication before use and is lower priority for initial modeling.

**Category Translation**

* Small lookup table with complete data quality.
* Can be merged with product information during transformation.


In [60]:


mql = pd.read_csv("olist_marketing_qualified_leads_dataset.csv")
closed_deals = pd.read_csv("olist_closed_deals_dataset.csv")

marketing_datasets = {
    "marketing_qualified_leads": mql,
    "closed_deals": closed_deals
}

for name, df in marketing_datasets.items():
    
    print("="*60)
    print(f"DATASET: {name}")
    print("="*60)

    print(f"\nRows: {df.shape[0]:,}")
    print(f"Columns: {df.shape[1]}")

    print("\nColumn Names:")
    print(df.columns.tolist())

    print("\nMissing Values:")
    print(df.isnull().sum())

    print("\nDuplicate Rows:")
    print(df.duplicated().sum())

    print("\nData Types:")
    print(df.dtypes)

    print("\nSample Records:")
    print(df.head())

    print("\n")

DATASET: marketing_qualified_leads

Rows: 8,000
Columns: 4

Column Names:
['mql_id', 'first_contact_date', 'landing_page_id', 'origin']

Missing Values:
mql_id                 0
first_contact_date     0
landing_page_id        0
origin                60
dtype: int64

Duplicate Rows:
0

Data Types:
mql_id                str
first_contact_date    str
landing_page_id       str
origin                str
dtype: object

Sample Records:
                             mql_id first_contact_date  \
0  dac32acd4db4c29c230538b72f8dd87d         2018-02-01   
1  8c18d1de7f67e60dbd64e3c07d7e9d5d         2017-10-20   
2  b4bc852d233dfefc5131f593b538befa         2018-03-22   
3  6be030b81c75970747525b843c1ef4f8         2018-01-22   
4  5420aad7fec3549a85876ba1c529bd84         2018-02-21   

                    landing_page_id          origin  
0  88740e65d5d6b056e0cda098e1ea6313          social  
1  007f9098284a86ee80ddeb25d53e0af8     paid_search  
2  a7982125ff7aa3b2054c6e44f9d28522  organic_search  
3 

### Marketing Qualified Leads (MQL) - Key Observations

* The dataset contains 8,000 marketing-qualified leads and 4 columns.
* mql_id appears to uniquely identify each lead.
* No duplicate records were found.
* Only 60 records have missing values in the origin column, representing less than 1% of the dataset.
* The dataset captures lead acquisition information, including first contact date, landing page, and acquisition source.
* Overall data quality is high and suitable for marketing funnel analysis.


### Closed Deals - Key Observations

* The dataset contains 842 successfully converted leads and 14 columns.
* mql_id can be used to link converted leads back to the marketing-qualified leads dataset.
* No duplicate records were detected.
* Several seller profiling fields contain substantial missing values, including company information, GTIN availability, stock levels, and product catalog size.
* Business segment, lead type, and business type are largely populated and remain useful for segmentation analysis.
* Despite missing seller attributes, the dataset is suitable for conversion, sales funnel, and lead quality analysis.


In [63]:
print("MQL Unique IDs:", mql['mql_id'].nunique())
print("Closed Deal Unique IDs:", closed_deals['mql_id'].nunique())

print(
    "Conversion Rate:",
    round(
        closed_deals['mql_id'].nunique() /
        mql['mql_id'].nunique() * 100,
        2
    ),
    "%"
)

MQL Unique IDs: 8000
Closed Deal Unique IDs: 842
Conversion Rate: 10.53 %


### Marketing Funnel Insights

* The Marketing Qualified Leads (MQL) dataset contains 8,000 leads generated through various acquisition channels.
* The Closed Deals dataset contains 842 successfully converted leads.
* Based on the available data, the lead-to-deal conversion rate is approximately 10.53%.
* This indicates that roughly 1 out of every 10 qualified leads ultimately became a seller on the platform.
* The datasets can be linked using mql_id, enabling future analysis of lead sources, sales performance, and conversion effectiveness.


### Marketing Funnel Summary

* The marketing dataset introduces a separate lead acquisition and conversion process that is independent from the marketplace transaction data.
* Out of 8,000 marketing-qualified leads, only 842 resulted in closed deals.
* This suggests an approximate lead-to-deal conversion rate of 10.5%.
* The datasets can be linked through mql_id to analyze lead sources, conversion performance, and sales effectiveness.
* These tables enable future marketing and sales funnel analytics beyond standard e-commerce reporting.
